### PR TITLE
refactor(mercury): update memory ordering for atomic operations

### PR DESCRIPTION
**Pull request by el-ev**
Sequencially consistent ordering was way too strong for the use cases. Using `Acquire` and `Release` is sufficient for our synchronization needs.

Changes:

* [`mercury/src/internal/pack/cache.rs`](diffhunk://#diff-fbec69ab7ba7f099b2d9456691864bdb1e6b7675ae742dfa4fd09c3c65d8244bL213-R213): Changed the memory ordering for `complete_signal.store` from `SeqCst` to `Release` in the `clear` method.

* `mercury/src/internal/pack/cache_object.rs`: 
  - Changed the memory ordering for `mem_recorder.fetch_sub` from `SeqCst` to `Release` in the `drop` method of `CacheObject`.
  - Changed the memory ordering for `mem_recorder.fetch_add` from `SeqCst` to `Release` in the `record_mem_size` method of `CacheObject`.
  - Changed the memory ordering for `CACHE_OBJS_MEM_SIZE.load` from `SeqCst` to `Acquire` in a commented-out section of `CacheObject`.
  - Changed the memory ordering for `complete_signal.load` from `SeqCst` to `Acquire` in the `drop` method of `ArcWrapper`. [[1]](diffhunk://#diff-be3c5031d6da12e5a85120085e6cfc668688e2146c2160202c1af517104822f6L300-R300) [[2]](diffhunk://#diff-be3c5031d6da12e5a85120085e6cfc668688e2146c2160202c1af517104822f6L313-R313)
  - Changed the memory ordering for `shared_flag.store` from `SeqCst` to `Release` in a test case.

* `mercury/src/internal/pack/decode.rs`: 
  - Changed the memory ordering for `cache_objs_mem.load` from **`Relaxed`** to `Acquire` in the `cache_objs_mem_used` method. 
  *On the contrary, this was too weak before.*
  - Changed the memory ordering for `count_c.store` from `Relaxed` to `Release` and `count.load` from `Relaxed` to `Acquire` in a test case.

### CODE REVIEW

File: mercury/src/internal/pack/cache_object.rs [hunk 1]
```diff
@@ -158,7 +158,7 @@ impl MemSizeRecorder for CacheObject {
     /// <br> since that, DO NOT modify `CacheObj` after recording
     fn record_mem_size(&self) {
         if let Some(mem_recorder) = &self.mem_recorder {
-            mem_recorder.fetch_add(self.mem_size(), Ordering::SeqCst);
+            mem_recorder.fetch_add(self.mem_size(), Ordering::Release);
```
Comment (MrBeanCpp) [inline mercury/src/internal/pack/cache_object.rs:161]: maybe `Ordering::AcqRel`?

File: mercury/src/internal/pack/decode.rs [hunk 2]
```diff
@@ -750,10 +750,10 @@ mod tests {
                 cnt += 1; //use entry here
             }
             tracing::info!("Received: {}", cnt);
-            count_c.store(cnt, Ordering::Relaxed);
+            count_c.store(cnt, Ordering::Release);
         }).await.unwrap();
         let p = pack.await.unwrap();
-        assert_eq!(count.load(Ordering::Relaxed), p.number);
+        assert_eq!(count.load(Ordering::Acquire), p.number);
```
Comment (MrBeanCpp) [inline mercury/src/internal/pack/decode.rs:756]: Here, `Ordering::Relaxed` maybe okay? `await` is used, so the sub-thread (`spawn_blocking`) has already over when `count.load()`.

**Review by MrBeanCpp — CHANGES_REQUESTED** [review]

File: mercury/src/internal/pack/decode.rs (hunk 2 above)
Comment (el-ev) [reply mercury/src/internal/pack/decode.rs:756]: Indeed. But it's in a test and only called once, don't be so serious :)

File: mercury/src/internal/pack/cache_object.rs (hunk 1 above)
Comment (el-ev) [reply mercury/src/internal/pack/cache_object.rs:161]: hmm... The reason why it is Release here, paired with Acquire in cache_objs_mem_used(), is the main thread will rely on the data changed here to decide whether to block the waitlist. But the other threads don't rely on the data modified by each other, so they can be non-Acquire.

Comment (MrBeanCpp) [reply mercury/src/internal/pack/cache_object.rs:161]: Sounds reasonable. hhh

**Review by MrBeanCpp — APPROVED** [review]

**Review by genedna — APPROVED** [review]

**Pull request merged**